web/src/main/java/org/springframework/security/web/savedrequest/CookieRequestCache.java

```diff
@@ -115,6 +136,51 @@ private static String decodeCookie(String encodedCookieValue) {
 		return new String(Base64.getDecoder().decode(encodedCookieValue.getBytes()));
 	}
 
+	private static String getCookiePath(HttpServletRequest request) {
+		final String contextPath = request.getContextPath();
+		if (StringUtils.isEmpty(contextPath)) {
+			return "/";
+		}
+		return contextPath;
+	}
+
+	private static String getServletPath(UriComponents uriComponents, String contextPath) {
+		String path = uriComponents.getPath();
+		if (path == null) {
+			return "";
+		}
+
+		if (StringUtils.isEmpty(contextPath) || contextPath.equals("/")) {
+			return path;
+		}
+
+		path = path.substring(contextPath.length());
+
+		return path;
+	}
+
+	private static String getRequestURL(String originalURI, UriComponents uriComponents) {
+		final String query = uriComponents.getQuery();
+		if (StringUtils.isEmpty(query)) {
+			return originalURI;
+		}
+
+		originalURI = originalURI.substring(0, originalURI.indexOf(query));
+		return StringUtils.trimTrailingCharacter(originalURI, '?');
+	}
+
+	private boolean matchesSavedRequest(HttpServletRequest request, SavedRequest savedRequest) {
+		if (savedRequest == null) {
+			return false;
+		} else if (savedRequest instanceof DefaultSavedRequest) {
```

Comment:
This `else if` block can be removed. We only have information about the saved request URL, so that is all we can really compare.

Reply:
I fixed these issues, please check.